client/client.go

```diff
@@ -163,44 +167,86 @@ func makeURL(base string, parts []string) string {
 	return base + strings.Join(parts, "/")
 }
 
-func (c *client) MakeServiceURL(serviceType string, parts []string) (string, error) {
+func (c *client) MakeServiceURL(serviceType, apiVersion string, parts []string) (string, error) {
 	return makeURL(c.baseURL, parts), nil
 }
 
 func (c *authenticatingClient) SetRequiredServiceTypes(requiredServiceTypes []string) {
 	c.requiredServiceTypes = requiredServiceTypes
 }
 
-func (c *authenticatingClient) SendRequest(method, svcType, apiCall string, requestData *goosehttp.RequestData) (err error) {
-	err = c.sendAuthRequest(method, svcType, apiCall, requestData)
+func (c *authenticatingClient) SendRequest(
+	method, svcType, apiVersion, apiCall string,
+	requestData *goosehttp.RequestData,
+) (err error) {
+	err = c.sendAuthRequest(method, svcType, apiVersion, apiCall, requestData)
 	if gooseerrors.IsUnauthorised(err) {
 		c.setToken("")
-		err = c.sendAuthRequest(method, svcType, apiCall, requestData)
+		err = c.sendAuthRequest(method, svcType, apiVersion, apiCall, requestData)
 	}
 	return
 }
 
-func (c *authenticatingClient) sendAuthRequest(method, svcType, apiCall string, requestData *goosehttp.RequestData) (err error) {
+func (c *authenticatingClient) sendAuthRequest(
+	method, svcType, apiVersion, apiCall string,
+	requestData *goosehttp.RequestData,
+) (err error) {
 	if err = c.Authenticate(); err != nil {
 		return
 	}
 
-	url, err := c.MakeServiceURL(svcType, []string{apiCall})
+	url, err := c.MakeServiceURL(svcType, apiVersion, []string{apiCall})
 	if err != nil {
 		return
 	}
 	return c.sendRequest(method, url, c.Token(), requestData)
 }
 
-func (c *authenticatingClient) MakeServiceURL(serviceType string, parts []string) (string, error) {
+// MakeServiceURL uses an endpoint matching the apiVersion for the given service type.
+// Given a major version only, the version with the highest minor will be used.
+// object-store and container service types have no versions
+func (c *authenticatingClient) MakeServiceURL(serviceType, apiVersion string, parts []string) (string, error) {
 	if !c.IsAuthenticated() {
 		return "", errors.New("cannot get endpoint URL without being authenticated")
 	}
-	url, ok := c.serviceURLs[serviceType]
+	serviceURL, ok := c.serviceURLs[serviceType]
 	if !ok {
 		return "", errors.New("no endpoints known for service type: " + serviceType)
 	}
-	return makeURL(url, parts), nil
+	requestedVersion, verErr := parseVersion(apiVersion)
+	if serviceType == "object-store" {
+		if apiVersion == "" || (verErr == nil && requestedVersion.major == 1) {
```

Comment:
Oh, I wasn't thinking of this approach. I thought getAPIVersions() would be modified to see if the URL was for object-store and just short circuit without calling the api and return an apiURLVersionInfo with 1.0 URL in it. That to me is the better place to put the short circuit, as it keeps the special case code next to the api in question. The code here is also a lot messier as there are checks for requestedVersion.major == 1 etc which would be avoided if the short circuit were where I am suggesting.
The fact that this works is good, let's not block on that, but my preference would be for the approach I suggest here.